File: OTEAServer/Controllers/ProvincesController.cs
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;
using System.Linq.Expressions;
using System.Text.Json;
using static OTEAServer.Controllers.CitiesController;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller class for provinces operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    [ApiController]
    [Route("Provinces")]
    public class ProvincesController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public ProvincesController(DatabaseContext context)
        {
            _context=context;
        }

        private class ProvinceDto
        {
            public int IdProvince { get; set; }
            public string Name { get; set; }
        }
        /// <summary>
        /// Method that obtains all provinces of a region
        /// </summary>
        /// <param name="idRegion">Region identifier</param>
        /// <param name="idCountry">Country identifier</param>
        /// <returns>Provinces list</returns>
        [HttpGet("allByRegion")]
        public IActionResult GetAllByRegion([FromQuery] int idRegion, [FromQuery] string idCountry, [FromQuery] string language)
        {
            try
            {
                IQueryable<Province> query = _context.Provinces
                    .Where(p => p.idRegion == idRegion && p.idCountry == idCountry)
                    .AsQueryable();

                Expression<Func<Province, ProvinceDto>> selector = p => new ProvinceDto
                {
                    IdProvince = p.idProvince,
                    Name = p.nameEnglish // Default value
                };
                switch (language)
                {
                    case "es":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameSpanish
                        };
                        query = query.OrderBy(p => p.nameSpanish); break;
                    case "fr":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameFrench
                        };
                        query = query.OrderBy(p => p.nameFrench); break;
                    case "eu":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameBasque
                        };
                        query = query.OrderBy(p => p.nameBasque); break;
                    case "ca":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameCatalan
                        };
                        query = query.OrderBy(p => p.nameCatalan); break;
                    case "gl":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameGalician
                        };
                        query = query.OrderBy(p => p.nameGalician); break;
                    case "pt":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.namePortuguese
                        };
                        query = query.OrderBy(p => p.namePortuguese); break;
                    case "de":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameGerman
                        };
                        query = query.OrderBy(p => p.nameGerman); break;
                    case "it":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameItalian
                        };
                        query = query.OrderBy(p => p.nameItalian); break;
                    case "nl":
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameDutch
                        };
                        query = query.OrderBy(p => p.nameDutch); break;
                    default:
                        selector = p => new ProvinceDto
                        {
                            IdProvince = p.idProvince,
                            Name = p.nameEnglish // Default value
                        };
                        query = query.OrderBy(p => p.nameEnglish); break;
                }
                var provinces=query.Select(selector).ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (ProvinceDto province in provinces)
                {
                    String rg = "{\"idProvince\":\"" + province.IdProvince + "\"," +
                            "\"name\":\"" + province.Name + "\"}";
                    result.Add(JsonDocument.Parse(rg));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains a province from the database
        /// </summary>
        /// <param name="idProvince">Province identifier</param>
        /// <param name="idRegion">Region identifier</param>
        /// <param name="idCountry">Country identifier</param>
        /// <returns></returns>
        [HttpGet("get")]
        public ActionResult<Province> Get([FromQuery] int idProvince, [FromQuery] int idRegion, [FromQuery] string idCountry)
        {
            try
            {
                var province = _context.Provinces.FirstOrDefault(p => p.idProvince == idProvince && p.idRegion == idRegion && p.idCountry == idCountry);

                if (province == null)
                    return NotFound();

                return province;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

       
    }
}

File: connection/src/main/java/otea/connection/controller/ProvincesController.java
package otea.connection.controller;

import com.google.gson.JsonObject;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.organization.data.geo.City;
import cli.organization.data.geo.Province;
import cli.organization.data.geo.Province;
import misc.ListCallback;
import otea.connection.ConnectionClient;
import otea.connection.api.ProvincesApi;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Controller class for provinces operation
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class ProvincesController {

    /**Provinces api to connect to the server*/
    private static ProvincesApi api;

    /**Controller instance*/
    private static ProvincesController instance;

    /**Number of attempts*/
    private static int numAttempts=0;

    /**Class constructor*/
    private ProvincesController(){
        api= ConnectionClient.getInstance().getRetrofit().create(ProvincesApi.class);
    }

    /**
     * Method that obtains the singleton instance of the controller
     *
     * @return Controller instance
     * */
    public static ProvincesController getInstance(){
        if(instance==null){
            synchronized (ProvincesController.class){
                if(instance==null){
                    instance=new ProvincesController();
                }
            }
        }
        return instance;
    }


    /**Refresh API*/
    public static void refreshApi(){
        instance=new ProvincesController();
    }

    /**
     * Method that obtains a province from the database
     *
     * @param idProvince - Province identifier
     * @param idRegion - Region identifier
     * @param idCountry - Country identifier
     * @return Province if success, null if not
     * */
    public static Province GetProvince(int idProvince, int idRegion, String idCountry){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Province> callable = new Callable<Province>() {
            @Override
            public Province call() throws Exception {
                Call<Province> call = api.GetProvince(idProvince,idRegion,idCountry);
                Response<Province> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Province> future = executor.submit(callable);
            Province result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains all provinces of a region
     *
     * @param idCountry - Country identifier
     * @return Provinces list
     * */
    public static void GetProvincesByRegion(int idRegion, String idCountry, ListCallback callback){

        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<JsonObject>> callable = new Callable<List<JsonObject>>() {
            @Override
            public List<JsonObject> call() throws Exception {
                Call<List<JsonObject>> call = api.GetProvincesByRegion(idRegion,idCountry, Locale.getDefault().getLanguage());
                Response<List<JsonObject>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };

        executor.submit(()->{
            try {
                List<JsonObject> result = callable.call();
                callback.onSuccess(result);
            } catch (Exception e) {
                callback.onError(e.getCause().toString());
            } finally {
                executor.shutdown();
            }
        });
        /*try {
            Future<List<JsonObject>> future = executor.submit(callable);
            List<JsonObject> result = future.get();
            executor.shutdown();
            List<Province> provinces=new ArrayList<>();
            for(JsonObject reg:result){
                int idProvince=reg.getAsJsonPrimitive("idProvince").getAsInt();
                String nameSpanish=reg.getAsJsonPrimitive("nameSpanish").getAsString();
                String nameEnglish=reg.getAsJsonPrimitive("nameEnglish").getAsString();
                String nameFrench=reg.getAsJsonPrimitive("nameFrench").getAsString();
                String nameBasque=reg.getAsJsonPrimitive("nameBasque").getAsString();
                String nameCatalan=reg.getAsJsonPrimitive("nameCatalan").getAsString();
                String nameDutch=reg.getAsJsonPrimitive("nameDutch").getAsString();
                String nameGalician=reg.getAsJsonPrimitive("nameGalician").getAsString();
                String nameGerman=reg.getAsJsonPrimitive("nameGerman").getAsString();
                String nameItalian=reg.getAsJsonPrimitive("nameItalian").getAsString();
                String namePortuguese=reg.getAsJsonPrimitive("namePortuguese").getAsString();
                provinces.add(new Province(idProvince,idRegion,idCountry,nameSpanish,nameEnglish,nameFrench,nameBasque,nameCatalan,nameDutch,nameGalician,nameGerman,nameItalian,namePortuguese));
            }
            return provinces;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetProvincesByRegion(idRegion,idCountry);
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }*/

    }
}
